### PR TITLE
add Tunisia Holidays

### PR DESCRIPTION
**Pull request by mr-wolf-gb**
i think we need to use other packages to calculate Hijri holidays like this package [hijri-dates](https://github.com/GeniusTS/hijri-dates), when this package added, i will add the Hijri holidays

### CODE REVIEW

File: src/Countries/Tunisia.php [hunk 1]
```diff
@@ -0,0 +1,2266 @@
...
+use Carbon\CarbonImmutable;
+use RuntimeException;
+
+class Tunisia extends Country
+{
+    /**
+     * @var array<int, array<int, array{name: string,date: string,slug: string}>> $hijriHolidays
+     */
+    protected array $hijriHolidays = [
+        "1970" => [
```
Comment (Nielsvanpach) [inline src/Countries/Tunisia.php:14]: isn't a solution like Turkey possible here?

https://github.com/spatie/holidays/blob/main/src/Countries/Turkey.php#L11

Comment (gaiththewolf) [reply src/Countries/Tunisia.php:14]: It's not a good solution I know I found the turkey code approved by the repo team so I adapted the code with the other Hijri holidays

**Comment by gaiththewolf** [comment]
I think the team of repo want to use the minimum of required packages

File: src/Countries/Tunisia.php (hunk 1 above)
Comment (gaiththewolf) [reply src/Countries/Tunisia.php:14]: We can use an external hijri converter package

Comment (gaiththewolf) [reply src/Countries/Tunisia.php:14]: > isn't a solution like Turkey possible here?
> 
> https://github.com/spatie/holidays/blob/main/src/Countries/Turkey.php#L11

Tomorrow I will apply the same Turkey code with more holidays

**Comment by mr-wolf-gb** [comment]
@Nielsvanpach i think these changes not viewed 😅

File: src/Calendars/HijriCalendar.php [hunk 2]
```diff
@@ -0,0 +1,375 @@
+<?php
+
+namespace Spatie\Holidays\Calendars;
+
+use Carbon\CarbonImmutable;
+use RuntimeException;
+
+trait HijriCalendar
```
Comment (Nielsvanpach) [inline src/Calendars/HijriCalendar.php:8]: As I understand correctly, these dates are not the same for all Islamic countries?

In that case, this class can probably not be reused, so I think it's better to move them to the Tunisia class.

File: src/Countries/Tunisia.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+use Carbon\CarbonImmutable;
+use Spatie\Holidays\Calendars\HijriCalendar;
+
+class Tunisia extends Country
+{
+    use HijriCalendar;
+
+    /**
+     * @return string
+     */
```
Comment (Nielsvanpach) [inline src/Countries/Tunisia.php:14]: No need for these docblocks when the type is already specified. Only array structures should be typed.

File: src/Countries/Tunisia.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public function countryCode(): string
+    {
+        return 'tn';
+    }
+
+    /**
+     * @param int $year
+     * @return array|CarbonImmutable[]|string[]
+     */
```
Comment (Nielsvanpach) [inline src/Countries/Tunisia.php:23]:
```suggestion
```

File: src/Countries/Tunisia.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    protected function allHolidays(int $year): array
+    {
+        $revolutionHoliday = [];
+
+        if (2022 > $year and $year >= 2011) {
```
Comment (Nielsvanpach) [inline src/Countries/Tunisia.php:28]:
```suggestion
        if (2022 > $year && $year >= 2011) {
```

File: tests/Countries/TunisiaTest.php [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    expect($holidays)
+        ->toBeArray()
+        ->not()->toBeEmpty();
+
+    expect(formatDates($holidays))->toMatchSnapshot();
+});
+
+it('can calculate tunisian holidays 2025', function () {
+    CarbonImmutable::setTestNowAndTimezone('2025-01-01');
```
Comment (Nielsvanpach) [inline tests/Countries/TunisiaTest.php:21]:
```suggestion
    CarbonImmutable::setTestNow('2025-01-01');
```

File: tests/Countries/TunisiaTest.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Spatie\Holidays\Tests\Countries;
+
+use Carbon\CarbonImmutable;
+use Spatie\Holidays\Holidays;
+
+it('can calculate tunisian holidays 2024', function () {
+    CarbonImmutable::setTestNowAndTimezone('2024-01-01');
```
Comment (Nielsvanpach) [inline tests/Countries/TunisiaTest.php:9]:
```suggestion
    CarbonImmutable::setTestNow('2024-01-01');
```

**Review by Nielsvanpach — CHANGES_REQUESTED** [review]

File: src/Calendars/HijriCalendar.php (hunk 2 above)
Comment (mr-wolf-gb) [reply src/Calendars/HijriCalendar.php]: the difference generally in one day, but i changed it to Tunisia only and i removed the trait

**Comment by Nielsvanpach** [comment]
Can you have a look at the failing phpstan errors?

**Comment by mr-wolf-gb** [comment]
> Can you have a look at the failing phpstan errors?

Done

**Comment by mr-wolf-gb** [comment]
@Nielsvanpach any suggestions here !

File: tests/Countries/TunisiaTest.php (hunk 7 above)
Comment (mr-wolf-gb) [reply tests/Countries/TunisiaTest.php]: I think everything is OK

File: tests/.pest/snapshots/Countries/TunisiaTest/it_can_calculate_tunisia_holidays_with_data_set___2024____2024_.snap [hunk 8]
```diff
@@ -12,31 +12,19 @@
         "date": "2024-04-09"
     },
     {
-        "name": "Eid al-Fitr 1. Day",
-        "date": "2024-04-10"
-    },
-    {
-        "name": "Eid al-Fitr 2. Day",
-        "date": "2024-04-11"
+        "name": "Eid al-Fitr",
```
Comment (Nielsvanpach) [inline tests/.pest/snapshots/Countries/TunisiaTest/it_can_calculate_tunisia_holidays_with_data_set___2024____2024_.snap:15]: Have you verified the updated snapshots? Are they correct?

**Review by Nielsvanpach — COMMENTED** [review]
Looks good!

File: tests/.pest/snapshots/Countries/TunisiaTest/it_can_calculate_tunisia_holidays_with_data_set___2024____2024_.snap (hunk 8 above)
Comment (mr-wolf-gb) [reply tests/.pest/snapshots/Countries/TunisiaTest/it_can_calculate_tunisia_holidays_with_data_set___2024____2024_.snap]: i fixed it, it's translation problem

**Pull request merged**